src/components/AdminBroadCasting/index.tsx

```diff
@@ -0,0 +1,89 @@
+import { useState } from 'react';
+import Link from 'next/link';
+import { DndContext } from '@dnd-kit/core';
+import { Label } from 'components/Label';
+import { Droppable } from 'components/Droppable';
+import { Draggable } from 'components/Draggable';
+import { FutureListComponents } from 'components/FutureList';
+
+export const AdminBroadCasting = (): JSX.Element => {
+  const containers = ['A'];
+  const wraps = ['B'];
+  const [parent, setParent] = useState(null);
+
+  const draggableMarkup = (
+    <Draggable id="draggable">
+      <p className="text-base pb-5 mx-4">
+        HTMLにはポータルという便利な要素がある
+      </p>
+      <div className="flex justify-between items-center">
+        <div className="ml-4">👨‍💼 松平 ケン</div>
+      </div>
+    </Draggable>
+  );
+
+  return (
+    <div className="items-center w-3/4 mx-auto">
+      <div className="pt-12 pb-16">
+        <div className="flex justify-center mb-6">
+          <Label type={'broadcasting'} />
+        </div>
+        <div className="text-right mx-auto">
+          <Link href="/adminAfterBroadCast">
+            <a>
+              <span className="bg-blue-100 hover:bg-blue-200 text-blue-500 text-base py-4 px-8 rounded">
+                放送を終了する
+              </span>
+            </a>
+          </Link>
+        </div>
+        <h1 className="text-center text-4xl font-bold">第4回 エンジビアの泉</h1>
+      </div>
+      <DndContext onDragEnd={handleDragEnd}>
+        <div className="flex justify-between">
+          <div>
+            <div className="bg-gray-300 text-black text-sm py-3 px-36 rounded font-bold">
+              フューチャー前
+            </div>
+            <ul>
+              <Droppable>{parent === null ? draggableMarkup : null}</Droppable>
+            </ul>
+          </div>
+          <div>
+            <div className="bg-gray-300 text-black text-sm py-3 px-32 rounded font-bold">
+              フューチャー中
+            </div>
+            <ul>
+              {containers.map((id) => (
+                <Droppable key={id} id={id}>
+                  {parent === id ? draggableMarkup : <FutureListComponents />}
+                </Droppable>
+              ))}
+            </ul>
+          </div>
+          <div>
+            <div className="bg-gray-300 text-black text-sm py-3 px-32 rounded font-bold">
+              フューチャー後
+            </div>
+            <ul>
+              {wraps.map((id) => (
+                <Droppable key={id} id={id}>
+                  {parent === id ? (
+                    draggableMarkup
+                  ) : (
+                    <FutureListComponents type={'afterFeaturing'} />
+                  )}
+                </Droppable>
+              ))}
+            </ul>
+          </div>
+        </div>
+      </DndContext>
+    </div>
+  );
+
+  function handleDragEnd(event: any) {
```

Comment:
hooksに切り出す形のほうが責務が別れて良さそう

Reply:
あと、eventの型がanyだと何が来ても受け入れてしまうのでもう少し縛ったほうが安全です。
具体的にはDragEventを引っ張ってきて明示する形です。

```
// dnd-kit/coreからDragEndEventの型をimport
// ちなみにonDragEndの引数の型から見れる
import { DndContext, DragEndEvent } from '@dnd-kit/core';
...

export const AdminBroadCasting = (): JSX.Element => {
  ...
  // parentの型を明示してあげないと引き渡す値がわからないよーと言われるので定義
  const [parent, setParent] = useState<string | null>(null);
  ...

  // any → DragEndEvent
  function handleDragEnd(event: DragEndEvent) {
    const { over } = event;
    setParent(over ? over.id : null);
  }
```

Reply:
わかりやすい解説ありがとうございます！
こんな型があったんですね。ドキュメントもっとしっかり読むようにします。

Reply:
## 質問です。
```tsx
function handleDragEnd(event: DragEndEvent) {
    const { over } = event;
    setParent(over ? over.id : null);
  }
```

この部分をhooksに切り出そうとしたのですが、setParentがあるので

```tsx
  const [parent, setParent] = useState<string | null>(null);
```

この部分も切り出してやるとなるとエラーが出るのですが、根本的に自分の考えが何か間違っているのでしょうか？？
ヒントみたいなものをいただけたら幸いです。

Reply:
お。どんなエラーでました？

Reply:
## 行ったこと
/hooks/useHandleDragEnd.tsx
```tsx
import { useState } from 'react';
import { DragEndEvent } from '@dnd-kit/core';

export const useHandleDragEnd = () => {
  const [parent, setParent] = useState<string | null>(null);

  function handleDragEnd(event: DragEndEvent) {
    const { over } = event;
    setParent(over ? over.id : null);
  }

  return {
    parent,
    handleDragEnd,
  };
};
```
にして切り出して、

```tsx
<DndContext onDragEnd={useHandleDragEnd}>
```
に変更しました

## エラー内容
```
Error: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://reactjs.org/link/invalid-hook-call for tips about how to debug and fix this problem.
```
調べた感じは恐らく、

- [Invalid Hook Call Warning](https://ja.reactjs.org/warnings/invalid-hook-call-warning.html)
- [React Hooksのルールをよく理解しないとハマるエラー対処法](https://qiita.com/tatsumin0206/items/4e1076e2deedf20a9485)

一つ目のReactのドキュメントで、
**Hooks can only be called inside the body of a function component.（フックは関数コンポーネントの本体でしか呼び出せません）**
とあるので、
```tsx
const [parent, setParent] = useState<string | null>(null);
```
この部分で、エラーになっていると予想しました。
ここからどうやってエラーを解消するのか詰まっている現状です。

Reply:
おっと、なんだろう？
自分は

```
// hooks/useDragEnd.ts
import { useState } from 'react';
import { DragEndEvent } from '@dnd-kit/core';

type Return = {
  parent: string | null;
  handleDragEnd: (event: DragEndEvent) => void;
};

export const useDragEnd = (): Return => {
  const [parent, setParent] = useState<string | null>(null);

  const handleDragEnd = (event: DragEndEvent) => {
    const { over } = event;
    setParent(over ? over.id : null);
  };

  return { parent, handleDragEnd };
};
```

これでいけましたね。
`AdminBroadCasting/index.tsx` にparentが残っているとかないですよね？

Reply:
アドバイス通りにコードを書き直してみてやってみたらできました！
原因は作成したカスタムフックでreturnしたparent, handleDragEndをAdminBroadCasting/index.tsxで呼び出していないことでした😰